internal/server/boltdbstate/job.go

```diff
@@ -1709,6 +1724,135 @@ func (s *State) taskComplete(jobId string) error {
 	return nil
 }
 
+// pipelineComplete will look up the referenced job to see if it has a PipelineStep ref
+// associated with it, and if so, progress the pipelineRun state machine.
+func (s *State) pipelineComplete(jobId string) error {
+	job, err := s.JobById(jobId, nil)
+	if err != nil {
+		s.log.Error("error getting job by id", "job", jobId, "err", err)
+		return err
+	} else if job.Pipeline == nil {
+		s.log.Trace("job is not part of a pipeline", "job", jobId)
+		return nil
+	}
+
+	// grab the pipeline run
+	run, err := s.PipelineRunGet(&pb.Ref_Pipeline{
+		Ref: &pb.Ref_Pipeline_Id{
+			Id: &pb.Ref_PipelineId{
+				Id: job.Pipeline.Pipeline,
+			},
+		},
+	}, job.Pipeline.RunSequence)
+	if run != nil && len(run.Jobs) < 1 {
+		err = status.Errorf(codes.FailedPrecondition,
+			"no jobs queued for pipeline run %q", run,
+		)
+	}
+	if err != nil {
+		s.log.Error("failed to retrieve pipeline to complete", "job", job.Id, "pipeline", job.Pipeline.Pipeline, "run", job.Pipeline.RunSequence)
+		return err
+	}
+
+	if job.State == pb.Job_ERROR {
+		run.Status = pb.PipelineRun_FAILED
+	} else if job.State == pb.Job_SUCCESS {
+		// If job Id matches last job queued by pipeline
+		if job.Id == run.Jobs[len(run.Jobs)-1].Id {
+			run.Status = pb.PipelineRun_COMPLETED
+			s.log.Trace("pipeline run is complete", "job", job.Id, "pipeline", job.Pipeline.Pipeline, "run", run.Sequence)
+		}
+	}
+
+	// PipelineRunPut the new state
+	if err := s.PipelineRunPut(run); err != nil {
+		s.log.Error("failed to complete pipeline run", "job", job.Id, "pipeline", job.Pipeline.Pipeline, "run", job.Pipeline.RunSequence)
+		return err
+	}
+
+	return nil
+}
+
+// pipelineAck checks if the referenced job id has a PipelineStep ref associated with it,
+// and if so, progress the PipelineRun state machine.
+func (s *State) pipelineAck(jobId string) error {
+	job, err := s.JobById(jobId, nil)
+	if err != nil {
+		s.log.Error("error getting job by id", "job", jobId, "err", err)
+		return err
+	} else if job.Pipeline == nil {
+		s.log.Trace("job is not part of a pipeline", "job", jobId)
+		return nil
+	}
+
+	// grab pipeline run that triggered the job based on the PipelineTask Ref
+	run, err := s.PipelineRunGet(&pb.Ref_Pipeline{
+		Ref: &pb.Ref_Pipeline_Id{
+			Id: &pb.Ref_PipelineId{
+				Id: job.Pipeline.Pipeline,
+			},
+		},
+	}, job.Pipeline.RunSequence)
+	if run != nil && len(run.Jobs) < 1 {
+		err = status.Errorf(codes.FailedPrecondition,
+			"no jobs queued for pipeline run %q", run,
+		)
+	}
+	if err != nil {
+		s.log.Error("failed to retrieve pipeline to ack", "job", job.Id, "pipeline id", job.Pipeline.Pipeline, "run", job.Pipeline.RunSequence)
+		return err
+	}
+	// Update the new pipeline run state
+	run.Status = pb.PipelineRun_RUNNING
+	s.log.Trace("pipeline is running", "job", job.Id, "pipeline", job.Pipeline.Pipeline, "run", run.Sequence)
+	if err := s.PipelineRunPut(run); err != nil {
```

Comment:
This could be updated to be a no-op (i.e. skip the put all together) if the existing run we got from 1789 is already `RUNNNING` to save us a database write.

Reply:
This is addressed in https://github.com/hashicorp/waypoint/pull/3599/commits/ba543a95b1cb53fbb22822a4850d2ea972523ef2, but Github didn't pick it up as resolved